### DISCUSSION TITLE
LCD flipped

### DISCUSSION THREAD

**Discussion by vujnovich** [probonopd/MiniDexed#412] [Q&A]
good day. i made a frankenstein case with hot glue and a bunch of plastic... I got the lcd l2s with the backpack.... working fine, however, the thing is upside down. Is there anything i can throw in the ini to flip this thing?

**Comment by probonopd** [comment]
Hi @vujnovich I don't think it's currently supported, but you could ask in https://github.com/rsta2/circle/ just to be sure.

**Reply by vujnovich** [reply]
So I threw mt32 pi on the SD card to give it a go, and inside the ini file, it actually has an option to flip the lcd among many other things...  So if the base of minidexed is based off mt32pi , I believe it might be in. I can follow up and show a photo I snap of the option in the mt32 ini, but hopefully I can get a clear answer if we can get it going as is.    I suppose I can flip the thing over, however the buttons and encoders would than be above the display , just making it seem like a poor planning choice, since your hand would be covering the display when you're making selections and changing options!  🤣

**Reply by probonopd** [reply]
mt32pi and MiniDexed are both based on https://github.com/rsta2/circle/.

The code in question seems to be https://github.com/dwhinham/mt32-pi/blob/bf95cbb240a7ba2bf71e45ef74fb0eaa0dad2409/src/lcd/drivers/ssd1306.cpp#L181-L187.

mt32 driver: 
https://github.com/dwhinham/mt32-pi/blob/master/src/lcd/drivers/ssd1306.cpp

Circle driver:
https://github.com/rsta2/circle/blob/master/addon/display/ssd1306device.cpp

The functionality in question would need to be backported to the driver in Circle, then we could use it.

Reference:
* https://github.com/rsta2/circle/issues/342

**Reply by vujnovich** [reply]
Thanks for making that post. I do appreciate it! You're a good dude at the helm of the project.  I don't know anything about building and compiling, or even coding, but I'd say it'd be a simple thing to look at where the rotation option code is on mt32pi and just copy and paste it in. Of course it isn't that simple though.  Mt32pi surely has many features and more importantly, options like flipping lcd that helps setups for convenience.  Maybe once the lcd flipping is looked into and implemented, other stuff can come our way in the future...   Again, thanks for taking my request seriously!

**Reply by diyelectromusic** [reply]
No its not supported in circle.  The ssd1306 implementation in circle is just as a "character device" rather than a true graphics device, this means it reflects what can be done with the HD44780 at present and there is no mirroring or rotation.

It looks like it should be possible to have an alternative constructor to support rotation and mirroring on initialisation, similar to the graphical driver in the MT32-Pi, but I don't know what that means for the HD44780 which I don't think supports rotation/mirroring?  I'll have to have a look and see and then see what Rene thinks about it...

Kevin

**Comment by diyelectromusic** [comment]
I have a candidate version that does this, see https://github.com/diyelectromusic/MiniDexed/tree/ssd1306rotate

But this relies on a PR being accepted into the develop branch of circle first: https://github.com/rsta2/circle/pull/343

Watch this space...

Kevin

**Reply by probonopd** [reply]
You're the man! :+1:

**Reply by diyelectromusic** [reply]
BTW develop branch of circle has now been updated, so you could give this a try if you wish :)

Note: this is only possible for SSD1306 displays not HD44780, hence the suggested naming of the configuration options.  That could change in future of course if we end up supporting more displays!

Kevin

**Reply by probonopd** [reply]
Pull request opened: https://github.com/probonopd/MiniDexed/pull/419. Let's test that build.

**Reply by probonopd** [reply]
@vujnovich please test [MiniDexed_2023-01-21-63fac3c](https://github.com/probonopd/MiniDexed/suites/10502991436/artifacts/521205226) and please let us know in https://github.com/probonopd/MiniDexed/pull/422 whether it works for you. Son't forget to edit `SSD1306LCDRotate=1` in `minidexed.ini`. Thanks!

**Comment by vujnovich** [comment]
This is great. I'll surely update my pi and see how it turns out.   This is
just nuts. I thought development of minidexed slowed to a halt because
there hasn't been a single YouTube video about it seemingly over the past 6
months

**Reply by probonopd** [reply]
Would you like to make a YouTube video (even if short) of your MiniDexed build? Maybe mentioning that we added the rotation feature? That'd be cool :+1:

**Reply by Banana71** [reply]
@vujnovich  👁️‍🗨️I really want to see your Frankenstein case

**Reply by diyelectromusic** [reply]
Nah, we keep tinkering :)

Just no shouting about it (much)!

Kevin
